tensorflow/core/common_runtime/simple_placer.cc

```diff
@@ -785,9 +785,23 @@ Status SimplePlacer::Run() {
 
     // Heuristic A application.
     if (IsGeneratorNode(node)) {
-      const Node* output = (*node->out_edges().begin())->dst();
+      EdgeSet::const_iterator it(node->out_edges().begin());
+
+      const Node* output = (*it)->dst();
       const string& output_device_name = output->assigned_device_name();
-      if (CanAssignToDevice(output_device_name, devices)) {
+      it++;
+
+      bool all_consumers_on_same_device(true);
+      while (it != node->out_edges().end()) {
+        if ((*it)->dst()->assigned_device_name() != output_device_name) {
+          all_consumers_on_same_device = false;
+          break;
+        }
+        it++;
+      }
+
```

Comment:
My code pulls the name from the 1st iterator value, then tests it against all of the others.  Your code captures 'output_device_name', which is the name from the 1st iterator value.  This would need to be read in your version of the code too.

What we need is a:

std::when_calling_with_elements_from_a_collection_every_return_value_of_a_functor_is_the_same()

It might exist.  My in-depth knowledge of the std c++ library is limited to pre 2004.